V3 occasionally fails with message "SecretAccessKey contains special characters"

### Describe the bug

We use OIDC to assume roles into our AWS accounts. We've noticed that intermittently our runs get the following message:

```
Error: Could not assume role with OIDC: SecretAccessKey contains special characters.
```

A retry allows it to succeed - presumably based on the value that the secret access key AWS is generating has.

### Expected Behavior

Role to be assumed correctly.

### Current Behavior

```
Assuming role with OIDC
Assuming role with OIDC
Assuming role with OIDC
Assuming role with OIDC
Error: Could not assume role with OIDC: SecretAccessKey contains special characters.
```

### Reproduction Steps

Using v3 of the action, and the following config:

```yaml
    - id: assume-role
      name: "Assume role"
      uses: aws-actions/configure-aws-credentials@v3
      with:
        role-to-assume: "arn:aws:iam:ACCOUNT_ID:role/ROLE_NAME"
        aws-region: "REGION"  # we're using ap-southeast-2, but we see this also in tests for us-east-1
        mask-aws-account-id: "true"
```

### Possible Solution

AWS is generating a secret key that contains special characters, and V3 isn't handling it correctly

### Additional Information/Context

Full log (by "View raw log"):

```
2023-08-24T06:10:01.9880509Z ##[group]Run aws-actions/configure-aws-credentials@v3
2023-08-24T06:10:01.9880769Z with:
2023-08-24T06:10:01.9881157Z   role-to-assume: arn:aws:iam::XXXXXXXXXXXX:role/XYZ
2023-08-24T06:10:01.9881464Z   aws-region: ap-southeast-2
2023-08-24T06:10:01.9881705Z   role-duration-seconds: 3600
2023-08-24T06:10:01.9882012Z   role-session-name: GithubActions-github-actions-bot
2023-08-24T06:10:01.9882277Z   mask-aws-account-id: true
2023-08-24T06:10:01.9882493Z   audience: sts.amazonaws.com
2023-08-24T06:10:01.9882712Z env:
2023-08-24T06:10:01.9887131Z ##[endgroup]
2023-08-24T06:10:02.7748622Z Assuming role with OIDC
2023-08-24T06:10:02.8351796Z Assuming role with OIDC
2023-08-24T06:10:02.8845453Z Assuming role with OIDC
2023-08-24T06:10:03.0082267Z Assuming role with OIDC
2023-08-24T06:10:03.0714358Z Assuming role with OIDC
2023-08-24T06:10:03.1736220Z Assuming role with OIDC
2023-08-24T06:10:03.9789858Z Assuming role with OIDC
2023-08-24T06:10:05.9029889Z Assuming role with OIDC
2023-08-24T06:10:08.6040897Z Assuming role with OIDC
2023-08-24T06:10:14.5153224Z Assuming role with OIDC
2023-08-24T06:10:29.8006873Z Assuming role with OIDC
2023-08-24T06:10:54.8733099Z Assuming role with OIDC
2023-08-24T06:11:43.1612120Z ##[error]Could not assume role with OIDC: SecretAccessKey contains special characters.
```

--- a/README.md
+++ b/README.md
@@ -268,6 +268,15 @@ Your account ID is not masked by default in workflow logs since it's not conside
 #### Unset current credentials
 Sometimes, existing credentials in your runner can get in the way of the intended outcome, and the recommended solution is to include another step in your workflow which unsets the environment variables set by this action. Now if you set the `unset-current-credentials` input to `true`, the workaround is made eaiser
 
+#### Special characters in AWS_SECRET_ACCESS_KEY
+Some edge cases are unable to properly parse an `AWS_SECRET_ACCESS_KEY` if it
+contains special characters. For more information, please see the
+[AWS CLI documentation](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-troubleshooting.html#tshoot-signature-does-not-match).
+If you set the `special-characters-workaround` option, this action will
+continually retry fetching credentials until we get one that does not have
+special characters. This option overrides the `disable-retry` and
+`retry-max-attempts` options.
+
 ## OIDC
 
 We recommend using [GitHub's OIDC provider](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) to get short-lived AWS credentials needed for your actions. When using OIDC, this action will create a JWT unique to the workflow run, and it will use this JWT to assume the role. For this action to create the JWT, it is required for your workflow to have the `id-token: write` permission:

--- a/action.yml
+++ b/action.yml
@@ -70,6 +70,9 @@ inputs:
   retry-max-attempts:
     description: The maximum number of attempts it will attempt to retry the assume role call. By default it will retry 12 times
     required: false
+  special-characters-workaround:
+    description: Some environments do not support special characters in AWS_SECRET_ACCESS_KEY. This option will retry fetching credentials until the secret access key does not contain special characters. This option overrides disable-retry and retry-max-attempts. This option is disabled by default
+    required: false
 outputs:
   aws-account-id:
     description: The AWS account ID for the provided credentials

--- a/src/assumeRole.ts
+++ b/src/assumeRole.ts
@@ -5,7 +5,7 @@ import * as core from '@actions/core';
 import type { AssumeRoleCommandInput, STSClient, Tag } from '@aws-sdk/client-sts';
 import { AssumeRoleCommand, AssumeRoleWithWebIdentityCommand } from '@aws-sdk/client-sts';
 import type { CredentialsClient } from './CredentialsClient';
-import { errorMessage, isDefined, sanitizeGitHubVariables, verifyKeys } from './helpers';
+import { errorMessage, isDefined, sanitizeGitHubVariables } from './helpers';
 
 async function assumeRoleWithOIDC(params: AssumeRoleCommandInput, client: STSClient, webIdentityToken: string) {
   delete params.Tags;
@@ -17,7 +17,6 @@ async function assumeRoleWithOIDC(params: AssumeRoleCommandInput, client: STSCli
         WebIdentityToken: webIdentityToken,
       })
     );
-    verifyKeys(creds.Credentials);
     return creds;
   } catch (error) {
     throw new Error(`Could not assume role with OIDC: ${errorMessage(error)}`);
@@ -49,7 +48,6 @@ async function assumeRoleWithWebIdentityTokenFile(
         WebIdentityToken: webIdentityToken,
       })
     );
-    verifyKeys(creds.Credentials);
     return creds;
   } catch (error) {
     throw new Error(`Could not assume role with web identity token file: ${errorMessage(error)}`);
@@ -60,7 +58,6 @@ async function assumeRoleWithCredentials(params: AssumeRoleCommandInput, client:
   core.info('Assuming role with user credentials');
   try {
     const creds = await client.send(new AssumeRoleCommand({ ...params }));
-    verifyKeys(creds.Credentials);
     return creds;
   } catch (error) {
     throw new Error(`Could not assume role with user credentials: ${errorMessage(error)}`);

--- a/src/helpers.ts
+++ b/src/helpers.ts
@@ -93,18 +93,21 @@ export function reset() {
 
 export function verifyKeys(creds: Partial<Credentials> | undefined) {
   if (!creds) {
-    return;
+    return false;
   }
   if (creds.AccessKeyId) {
     if (SPECIAL_CHARS_REGEX.test(creds.AccessKeyId)) {
-      throw new Error('AccessKeyId contains special characters.');
+      core.debug('AccessKeyId contains special characters.');
+      return false;
     }
   }
   if (creds.SecretAccessKey) {
     if (SPECIAL_CHARS_REGEX.test(creds.SecretAccessKey)) {
-      throw new Error('SecretAccessKey contains special characters.');
+      core.debug('SecretAccessKey contains special characters.');
+      return false;
     }
   }
+  return true;
 }
 
 // Retries the promise with exponential backoff if the error isRetryable up to maxRetries time.

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,4 +1,5 @@
 import * as core from '@actions/core';
+import type { AssumeRoleCommandOutput } from '@aws-sdk/client-sts';
 import { assumeRole } from './assumeRole';
 import { CredentialsClient } from './CredentialsClient';
 import {
@@ -8,6 +9,7 @@ import {
   exportCredentials,
   exportAccountId,
   unsetCredentials,
+  verifyKeys,
 } from './helpers';
 
 const DEFAULT_ROLE_DURATION = 3600; // One hour (seconds)
@@ -43,10 +45,20 @@ export async function run() {
     const unsetCurrentCredentialsInput = core.getInput('unset-current-credentials', { required: false }) || 'false';
     const unsetCurrentCredentials = unsetCurrentCredentialsInput.toLowerCase() === 'true';
     const disableRetryInput = core.getInput('disable-retry', { required: false }) || 'false';
-    const disableRetry = disableRetryInput.toLowerCase() === 'true';
+    let disableRetry = disableRetryInput.toLowerCase() === 'true';
+    const specialCharacterWorkaroundInput =
+      core.getInput('special-characters-workaround', { required: false }) || 'false';
+    const specialCharacterWorkaround = specialCharacterWorkaroundInput.toLowerCase() === 'true';
     let maxRetries = parseInt(core.getInput('retry-max-attempts', { required: false })) || 12;
-    if (maxRetries < 1) {
-      maxRetries = 1;
+    switch (true) {
+      case specialCharacterWorkaround:
+        // 😳
+        maxRetries = Number.MAX_SAFE_INTEGER;
+        disableRetry = false;
+        break;
+      case maxRetries < 1:
+        maxRetries = 1;
+        break;
     }
     for (const managedSessionPolicy of managedSessionPoliciesInput) {
       managedSessionPolicies.push({ arn: managedSessionPolicy });
@@ -129,25 +141,30 @@ export async function run() {
 
     // Get role credentials if configured to do so
     if (roleToAssume) {
-      const roleCredentials = await retryAndBackoff(
-        async () => {
-          return assumeRole({
-            credentialsClient,
-            sourceAccountId,
-            roleToAssume,
-            roleExternalId,
-            roleDuration,
-            roleSessionName,
-            roleSkipSessionTagging,
-            webIdentityTokenFile,
-            webIdentityToken,
-            inlineSessionPolicy,
-            managedSessionPolicies,
-          });
-        },
-        !disableRetry,
-        maxRetries
-      );
+      let roleCredentials: AssumeRoleCommandOutput;
+      do {
+        // eslint-disable-next-line no-await-in-loop
+        roleCredentials = await retryAndBackoff(
+          async () => {
+            return assumeRole({
+              credentialsClient,
+              sourceAccountId,
+              roleToAssume,
+              roleExternalId,
+              roleDuration,
+              roleSessionName,
+              roleSkipSessionTagging,
+              webIdentityTokenFile,
+              webIdentityToken,
+              inlineSessionPolicy,
+              managedSessionPolicies,
+            });
+          },
+          !disableRetry,
+          maxRetries
+        );
+        // eslint-disable-next-line no-unmodified-loop-condition
+      } while (specialCharacterWorkaround && !verifyKeys(roleCredentials.Credentials));
       core.info(`Authenticated as assumedRoleId ${roleCredentials.AssumedRoleUser!.AssumedRoleId!}`);
       exportCredentials(roleCredentials.Credentials, outputCredentials);
       // We need to validate the credentials in 2 of our use-cases
@@ -173,6 +190,7 @@ export async function run() {
 }
 
 /* c8 ignore start */
+/* istanbul ignore next */
 if (require.main === module) {
   (async () => {
     await run();

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -556,61 +556,62 @@ describe('Configure AWS Credentials', () => {
     expect(mockedSTS.commandCalls(AssumeRoleWithWebIdentityCommand).length).toEqual(1);
   });
 
-  test('role assumption fails if access key id contains special characters', async () => {
-    jest.spyOn(core, 'getInput').mockImplementation(mockGetInput({ ...ASSUME_ROLE_INPUTS }));
+  test('special character workaround works for AWS_ACCESS_KEY_ID', async () => {
+    jest
+      .spyOn(core, 'getInput')
+      .mockImplementation(mockGetInput({ ...ASSUME_ROLE_INPUTS, 'special-characters-workaround': 'true' }));
 
-    mockedSTS.on(AssumeRoleCommand).resolves({
-      Credentials: {
-        AccessKeyId: 'asdf+',
-        SecretAccessKey: FAKE_STS_SECRET_ACCESS_KEY,
-        SessionToken: FAKE_STS_SESSION_TOKEN,
-        Expiration: new Date(8640000000000000),
-      },
-    });
+    mockedSTS
+      .on(AssumeRoleCommand)
+      .resolvesOnce({
+        Credentials: {
+          AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
+          SecretAccessKey: 'asdf+',
+          SessionToken: FAKE_STS_SESSION_TOKEN,
+          Expiration: new Date(8640000000000000),
+        },
+      })
+      .resolves({
+        Credentials: {
+          AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
+          SecretAccessKey: 'asdf',
+          SessionToken: FAKE_STS_SESSION_TOKEN,
+          Expiration: new Date(8640000000000000),
+        },
+      });
 
     await run();
 
-    expect(mockedSTS.commandCalls(AssumeRoleCommand).length).toEqual(12);
-    expect(core.setFailed).toHaveBeenCalledWith(
-      'Could not assume role with user credentials: AccessKeyId contains special characters.'
-    );
+    expect(mockedSTS.commandCalls(AssumeRoleCommand).length).toEqual(2);
   });
 
-  test('role assumption fails if secret access key contains special characters', async () => {
-    jest.spyOn(core, 'getInput').mockImplementation(mockGetInput({ ...ASSUME_ROLE_INPUTS }));
+  test('special character workaround works for AWS_SECRET_ACCESS_KEY', async () => {
+    jest
+      .spyOn(core, 'getInput')
+      .mockImplementation(mockGetInput({ ...ASSUME_ROLE_INPUTS, 'special-characters-workaround': 'true' }));
 
-    mockedSTS.on(AssumeRoleCommand).resolves({
-      Credentials: {
-        AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
-        SecretAccessKey: 'asdf+',
-        SessionToken: FAKE_STS_SESSION_TOKEN,
-        Expiration: new Date(8640000000000000),
-      },
-    });
-
-    await run();
-
-    expect(mockedSTS.commandCalls(AssumeRoleCommand).length).toEqual(12);
-    expect(core.setFailed).toHaveBeenCalledWith(
-      'Could not assume role with user credentials: SecretAccessKey contains special characters.'
-    );
-  });
-
-  test('role assumption succeeds if keys have no special characters', async () => {
-    jest.spyOn(core, 'getInput').mockImplementation(mockGetInput({ ...ASSUME_ROLE_INPUTS }));
-
-    mockedSTS.on(AssumeRoleCommand).resolves({
-      Credentials: {
-        AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
-        SecretAccessKey: FAKE_STS_SECRET_ACCESS_KEY,
-        SessionToken: FAKE_STS_SESSION_TOKEN,
-        Expiration: new Date(8640000000000000),
-      },
-    });
+    mockedSTS
+      .on(AssumeRoleCommand)
+      .resolvesOnce({
+        Credentials: {
+          AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
+          SecretAccessKey: 'asdf+',
+          SessionToken: FAKE_STS_SESSION_TOKEN,
+          Expiration: new Date(8640000000000000),
+        },
+      })
+      .resolves({
+        Credentials: {
+          AccessKeyId: FAKE_STS_ACCESS_KEY_ID,
+          SecretAccessKey: 'asdf',
+          SessionToken: FAKE_STS_SESSION_TOKEN,
+          Expiration: new Date(8640000000000000),
+        },
+      });
 
     await run();
 
-    expect(mockedSTS.commandCalls(AssumeRoleCommand).length).toEqual(1);
+    expect(mockedSTS.commandCalls(AssumeRoleCommand).length).toEqual(2);
   });
 
   test('max retries is configurable', async () => {

